blog-posts/optimze-gatsby-urls.md

```diff
@@ -0,0 +1,153 @@
+- [https://satellytes.com/career/325420-senior-frontend-engineer-(mwx)/index.html](https://satellytes.com/career/325420-senior-frontend-engineer-(mwx)/index.html)
+- all 3 links above, but starting with `https://www.`
+- all 3 links above, but starting with `http://`
+- all 3 links above, but starting with `http://www.`
+
+In general, it's good that all links are supported, as it lowers the possibility of a wrong URL entered by the user. But it also comes with 2 downsides:
+
+1. Degraded performance for the user as only one link returns the HTML file directly. All other links are redirects.
+2. Duplicate content for search engines. A page should only have one URL (or at least one "original" URL)
+
+## Optimize URLs for the user
+
+If you use Netlify for hosting, only one URL delivers the actual HTML file, and that's the URL with `https`, no `www.`  and a trailing slash. All other URLs will be redirected. In the worst case, you have 3 redirects before the user will get the HTML file:
+
+![images/how-to-optimize-gatsby-url-redirect.png](images/how-to-optimize-gatsby-url-redirect.png)
+
+1. `307` redirect from `http` to `https`
+2. `301` redirect from `www.` to plain URL
+3. `301` redirect from missing trailing slash to `/`
+4. `200` with the HTML document 🎉
+
+The URLs that the users are using are the ones we put on our website. That's why we should make sure to always use the direct URL. We did this by creating a general `Link` component, that we use for all links we put on our side. It looks something like this:
+
+```tsx
+import { Link as GatsbyLink } from 'gatsby';
+
+export const Link = (props: LinkProps): JSX.Element => {
+  const {
+    to,
+		children,
+		...rest
+  } = props;
+
+  const isExternalLink = to.toString().startsWith('http');
+  if (isExternalLink) {
+    return (
+      <a
+        href={to}
+        target="_blank"
+        rel="noopener"
+        {...rest}
+      >
+        {children}
+      </a>
+    );
+  }
+
+  // here we add the traling slash for every URL we put into a link
+  const toWithSlash = to.endsWith('/') ? to : `${to}/`;
+
+  return (
+    <GatsbyLink
+      to={toWithSlash}     
+      {...rest}
+    >
+      {children}
+    </GatsbyLink>
+  );
+};
+```
+
+> Check out the full component implementation on our public website repository: [https://github.com/satellytes/satellytes.com/blob/main/src/components/links/links.tsx](https://github.com/satellytes/satellytes.com/blob/main/src/components/links/links.tsx)
+
+With this component, we get consistency into our URLs (inconsistent URLs could come from authors or developers by accident). The trailing slash will be appended to all internal links and users don't have to deal with redirects - at least when they open the link in a new tab or share it to their friends, as client-side routing with `GatsbyLink` is not affected by this. Having all URLs aligned also helps us with SEO, which we go into detail in the next paragraph.
```

Comment:
In addition the first sentence is also difficult to understand with the parentheses. I proposed a slightly different sentence before, but you probably did not like it :D 

Here the same sentence with the addition about authoring wrong content. Maybe it helps to rephrase?

> This link component enforces a consistent format for any given URL if someone accidentally decided for a url without a trailing slash

Reply:
I removed the sentence in the parenthesis. I put the sentence into the paragraph above the code example:

> That's why we should make sure to always use the direct URL on our page, even if the developers or authors put in the wrong one.